### PR TITLE
Add lazperf pcpatch

### PR DESCRIPTION
**Pull request by pblottiere**
This PR adds a new kind of PCPATCH using the laz compression coming from lazperf : 
  - patch PCPATCH_LAZPERF
  - unit tests
  - install tests
  - doc
  - travis builds the lazperf library to run unit tests

See issue https://github.com/pgpointcloud/pointcloud/issues/81

### CODE REVIEW

**Comment by pramsey** [comment]
This is very exciting, can you share some experiences w/ real-world use? Compression ratios on non-synthetic LAS data? Performance of compress/decompress?

**Comment by pblottiere** [comment]
@pramsey I've put in place a simple benchmark to compare compressions ratios between "no compression", "dimensional compression" and "laz compression" in the directory tools/benchmark_compression/. This benchmark is currently running within the travis build and results are interesting : 

```
$ cd tools/benchmark_compression && sh compression_benchmark.sh

           Table            |  Size   | External Size 

----------------------------+---------+---------------

 pa_compression_none        | 1600 kB | 1584 kB

 pa_compression_dimensional | 760 kB  | 736 kB

 pa_compression_laz         | 128 kB  | 32 kB

 pointcloud_formats         | 16 kB   | 16 kB

(4 rows)
```

**Comment by pramsey** [comment]
50:1? Is that on synthetic data?

**Comment by pblottiere** [comment]
Benchmark is running against generated data :

```
INSERT INTO pa_compression_dimensional (pa)
SELECT PC_Patch(PC_MakePoint(5, ARRAY[x,y,z,intensity]))
FROM (
 SELECT
 -127+a/100.0 AS x,
   45+a/100.0 AS y,
        1.0*a AS z,
         a/10 AS intensity,
         a/400 AS gid
  FROM generate_series(1,100000) AS a
) AS values GROUP BY gid;
```

**Comment by hobu** [comment]
> 50:1? Is that on synthetic data?

Must be. I typically see 4:1 to 6:1 with real lazperf-encoded data that doesn't have a model like LAZ along with it. 10:1 to 12:1 with LAZ is possible, but it needs to be either really boring data, well sorted, or highly return-heavy (LAZ does a great job collapsing multi-return data together).

The short answer of why use lazperf is you are a leveraging your knowledge of the arrangement of the stream (high periodicity, high coherence) with the application of [arithmetic encoding](http://www.drdobbs.com/cpp/data-compression-with-arithmetic-encodin/240169251). lazperf by itself is just the application of the arithmetic encoder plus description of the periodic point fields. Something like LAZ takes that even further to model across fields and across points, but LAZ must assume a specific set of fields and also data content that behaves a certain way. We can get a long way with just application of the arithmetic encoder, which is what we're doing when we say "lazperf" in this fashion.

This PR made me go back and do some simulation, testing, and verification of lazperf in relation to lzma (one of the better generic compression libraries). I hope to have a write-up soon with some simulation and statistics. lazperf is likely to win in both space and time dimensions due to the fact that we are leveraging our knowledge about how the data are arranged to apply arithmetic encoding. Something generic doesn't get that advantage without making a bunch of heuristic passes or grinding over the data multiple times.

**Comment by pramsey** [comment]
Will, I'm tempted to just push the green button, but I don't feel like I have the "social license" to do that anymore. @strk OK w/ this pull?

**Comment by hobu** [comment]
I think this PR should wait until we can get lazperf built up into a Real Library. That's in progress right now because PDAL is moving to using lazperf exclusively for LAZ and compression support. My other hesitation is the implementation is pretty verbose, and that is the fault of lazperf not providing much convenience. I'd like to improve that story so there's less code to maintain and duplicate for everyone using the library.

**Comment by pramsey** [comment]
I noticed some CPP/ shims in there. Would lasperf the Library include a C API?

**Comment by hobu** [comment]
> Would lasperf the Library include a C API?

We weren't planning on developing one, but we'd happily merge one if it showed up 😄

**Comment by strk** [comment]
@pramsey for what is worth, you have my full trust in accepting
contributions into your own creature :)

**Comment by pramsey** [comment]
I feel a long distance from it at this point :) It's been quite a while since I've changed a substantial line.

**Comment by pramsey** [comment]
Oh @hobu, approximate ETA on when @pblottiere should look at revising this against a new and improved lasperf?

**Comment by hobu** [comment]
> approximate ETA on when @pblottiere should look at revising this against a new and improved lasperf?

I don't have one, so maybe just merge it and deal with the consequences later ;) We're all kind of waiting on what @rapidlasso does for LAZ 1.4 before acting anyway.

**Comment by rapidlasso** [comment]
Things will start happening mid-June. Funding is (hopefully) lined-up (and not falling through) and time is reserved. Discussions will be via the LAS room (http://groups.google.com/group/lasroom) ...

**Comment by hobu** [comment]
I ran some tests of lazperf (LAZ without the model) versus [LZMA](http://tukaani.org/xz/) to verify that yes indeed, it is a good thing to do. The results confirmed my expectations when run across 3500+ linear mode LiDAR files.

Compression sizes of LZMA vs Lazperf were comparable, which is expected due them applying a similar compression technique to the data. The banding in the data are various LAS point formats and groups of files that do or don't have actual data in those dimensions. Some more investigation might be interesting, but I was just looking for confirmation. 
laz-vs-lzma-size

Compression timing on the other hand, is a drastic win for Lazperf. This is also expected. A generic algorithm like LZMA is computing statistics on the the data is it goes, which converges will in the space dimension at the cost of the time one. A quick glance at the graph shows it is consistently 8x faster to compress the data. I didn't do a decompression test, but I would expect that both are quick and similar.
laz-vs-lzma-time

Given the choice if your model fits, you want to use LASzip, however. Whereas Lazperf is the arithmetic encoder minus the data model, LAZ gets great leverage from this model for significant storage savings. 
laszip-vs-lazperf-size
laszip-vs-lzma-size

To summarize, Lazperf is a very good choice for generically formatted, periodic point cloud data. Use LAZ if your data fit that arrangement, but if not (extra dimensions, different fields), Lazperf does better than any generic compression algorithm due to you prompting it with the expected storage fields when you construct both the compressor and decompressor. 

Combined with JavaScript capabilities of being able to compile Lazperf to JavaScript with Emscripten (used in http://potree.org and http://plas.io), it makes a compelling combination. Additionally, we are using the same technique for the [Greyhound](https://github.com/hobu/greyhound) and [Entwine](https://github.com/connormanning/entwine) software combination (point cloud web service access).

**Comment by pblottiere** [comment]
@hobu Very interesting results! It would be great if we could capitalize this knowledge "somewhere".

**Comment by pblottiere** [comment]
> I don't have one, so maybe just merge it and deal with the consequences later ;) We're all kind of waiting on what rapidlasso does for LAZ 1.4 before acting anyway.
	
@pramsey Just out of curiosity, do you finally want to wait for the lazperf library to merge this PR? By the way, do not hesitate to give me your feelings on the code! I'll correct if you find something wrong.

**Comment by pramsey** [comment]
Functionally the code is fine, and I appreciate the addition of appropriate unit and integration tests. I'll make two notes, neither of which requires action, because they are so trivial and stupid:

* One thing I learned from submitting a patch to the PgSQL community was the wisdom of the directive: don't make white-space changes, ever. For things like patch review, having a patch that is truly nothing except the functional changes is very valuable. I takes a certain amount of discipline to avoid touching the little things around the edges that need fixing while working on your main patch, but it makes it easier to review, and thus easier to get accepted.
* In a similar vein, following the code standards of the existing project is a good idea. I see you've done that in existing files, but in your net new files you've used your preferred indentation. Consistency is the hobgoblin of my little mind.

**Comment by hobu** [comment]
👍 to merge from me as well. I agree with @pramsey's comment on format and style, but it's not my project 😄 

I expect there will be some changes related to lazperf, but this gets the ball rolling.

**Comment by pblottiere** [comment]
> In a similar vein, following the code standards of the existing project is a good idea. I see you've done that in existing files, but in your net new files you've used your preferred indentation. Consistency is the hobgoblin of my little mind.

Yes you're perfectly right. I'm gonna fix it tomorrow!

**Comment by pblottiere** [comment]
@pramsey I fixed format and style :sweat_smile:

**Comment by pramsey** [comment]
OK, you're merged with the big green button... that means you're now on the hook for lasperf support :)

**Comment by strk** [comment]
Please add a note in the NEWS file, if it wasn't in the PR already.
Format is the usual one ...

**Comment by strk** [comment]
On the README.md file:  "There are currently three supported compressions:" -- but now they are four ...
